flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -949,11 +982,73 @@ private void removeSlot(SlotID slotId) {
 					oldAllocationId,
 					new FlinkException("The assigned slot " + slot.getSlotId() + " was removed."));
 			}
+
+			removeSlotFromAllocatedResources(slotId);
 		} else {
 			LOG.debug("There was no slot registered with slot id {}.", slotId);
 		}
 	}
 
+	private void recordAllocatedSlotAndResource(TaskManagerSlot slot, ResourceProfile resourceProfile) {
+		ResourceID resourceID = slot.getSlotId().getResourceID();
+
+		if (!taskExecutorAllocationSlotsResource.containsKey(resourceID)) {
+			taskExecutorAllocationSlotsResource.put(resourceID,
+				Tuple2.of(new HashMap<>(), taskExecutorResources.get(resourceID)));
+		}
+		Tuple2<Map<SlotID, ResourceProfile>, ResourceProfile> allocatedSlotsResource =
+			taskExecutorAllocationSlotsResource.get(resourceID);
+
+		if (resourceProfile == ResourceProfile.UNKNOWN) {
+			resourceProfile = slot.getResourceProfile();
+		}
+		allocatedSlotsResource.f0.put(slot.getSlotId(), resourceProfile);
+		allocatedSlotsResource.f1 = allocatedSlotsResource.f1.minus(resourceProfile);
+	}
+
+	private void removeSlotFromAllocatedResources(SlotID slotID) {
+		ResourceID resourceID = slotID.getResourceID();
+		if (taskExecutorAllocationSlotsResource.containsKey(resourceID)) {
+			Tuple2<Map<SlotID, ResourceProfile>, ResourceProfile> allocatedSlotsResource =
+				taskExecutorAllocationSlotsResource.get(slotID.getResourceID());
+
+			ResourceProfile resourceProfile = allocatedSlotsResource.f0.remove(slotID);
+			if (resourceProfile != null) {
+				allocatedSlotsResource.f1 = allocatedSlotsResource.f1.merge(resourceProfile);
+			}
+
+			if (allocatedSlotsResource.f0.isEmpty()) {
+				taskExecutorAllocationSlotsResource.remove(resourceID);
+			}
+		}
+	}
+
+	private boolean hasEnoughResource(TaskManagerSlot targetSlot, ResourceProfile requiredResource) {
+		ResourceProfile targetSlotResource = targetSlot.getResourceProfile();
+		if (requiredResource == ResourceProfile.UNKNOWN) {
+			if (targetSlotResource == ResourceProfile.ANY) {
+				return true;
+			}
+			requiredResource = targetSlotResource;
+		}
+
+		ResourceID resourceID = targetSlot.getSlotId().getResourceID();
+		Tuple2<Map<SlotID, ResourceProfile>, ResourceProfile> allocatedResources =
+			taskExecutorAllocationSlotsResource.get(resourceID);
+		ResourceProfile remain = (allocatedResources == null) ?
+			taskExecutorResources.get(resourceID) : allocatedResources.f1;
+
+		// for now, we only do dynamic resource allocation for managed memory
+		return new ResourceProfile(
+			targetSlotResource.getCpuCores(),
```

Comment:
I think here might be `requiredSlotResource` instead of `targetSlotResource`, since the `targetSlotResource` may be less than the required one but there are still remaining resources.